Enhance documentation for using performance libraries together (#1137)
Shopify/quilt: packages/react-performance/README.md

The most basic way to use the tools in this package is to record information and display it locally to the user. In practice you usually only want to do this in development so that developers can easily see performance information.

> ### Adding some `<PerformanceMark />`s

Reviewer: ### Adding metric tracking using `<PerformanceMark />`

> We will need to add some `<PerformanceMark />` components to our React application in order to help `@shopify/react-performance` determine when your application is fully loaded and React has mounted your components.

Reviewer: ```suggestion
We will need to add some `<PerformanceMark stage="complete"/>` components to our React application in order to help `@shopify/react-performance` determine when your application is fully loaded and React has mounted your components.
```

Reviewer: I suggest this since `when your application is fully loaded and React has mounted your components` is only true with `stage="complete"`

```tsx
//Cats.tsx
import React from 'react';
import {PerformanceMark} from '@shopify/react-performance';

/**
 * Represents an entire page of our application

>  */

Reviewer: I wonder if this should be

```
export function Cats() {
  return (
    <>
      <section>
        <h1>Cats</h1>
        <ul>
          <li>Kokusho</li>
          <li>Moe</li>
          <li>Smoky</li>
        </ul>
      </section>
      <PerformanceMark stage="complete" id="cats" />
    </>
  );
```

so the complete will include the render of `<section />` as well. (have I been doing this all wrong actually?)

Author: I don't think it matters too much, but possibly slightly more accurate

export function Cats() {
  return (
    <section>
      <h1>Cats</h1>
      <ul>
        <li>Kokusho</li>
        <li>Moe</li>
        <li>Smoky</li>
      </ul>
      <PerformanceMark stage="complete" id="cats" />
    </section>
  );
}
```

> In production apps, it is important to always have one `<PerformanceMark stage="complete" />` on each page level component. Otherwise, client-side navigations may not be recorded properly.

Reviewer: ```suggestion
In production apps, it is important to always have one `<PerformanceMark stage="complete" />` on each page level component. Otherwise, client-side navigations may not be recorded properly. You should not use `usePerformanceMark("complete", "id")` in place of the component because...
```

Reviewer: The `...` is mostly in reference to a question I asked in a previous PR on why we dont use `usePerformanceMark` , I think the reason why is a hook version will get involved before the entire component finished rendering?

Author: I'll add a tiny bit more context here, but it feels more like going super deep on when and where you might miss performance marks and the ramifications belongs more in an "FAQ" or something.

Author: Also
> Where does Web include performance marks that aren't just in the pages?

From what I can see it literally *only* uses `usable` and `complete` (mostly `complete`, and our types currently disallow custom stages) and  I *think* only on page level components.

Reviewer 2: https://github.com/Shopify/web/blob/master/app/foundation/ErrorBoundary/ErrorBoundary.tsx#L55 and probably should typically put one in the `NotFound` component to be safe

Reviewer: so I didnt add this in online-store-web yet but I do kind of wonder. Would this also make sense to add for an embedded modal? or just a large modal?

> 

Reviewer: Maybe we need to add an example for `<PerformanceMark stage="usable"/>`

Author: We have that below, titled `### Tracking _usable_ vs _complete_`

Reviewer: Yes, please ignore this comment, I realized that later on.

> #### Displaying data using `useNavigationListener`

Reviewer: ```suggestion
### Displaying real time navigation data using `useNavigationListener`
```

import React, {useState} from 'react';
import {useNavigationListener, Navigation} from '@shopify/react-performance';

/**
 * A component which displays simple performance stats about the last navigation performed by the user

>  */

Reviewer: ❤️

### Tracking _usable_ vs _complete_

In a more complex application with multiple pages and calls for remote data, it is helpful to use multiple `<PerformanceMark />` components to track data for when the page is _usable_ vs _complete_. As a rule of thumb, consider a page _usable_ when the top-level React component for it has mounted, and _complete_ when all asynchronous work (such as data fetching) required for rendering the page content has been completed.

For example, we can update our `<Cats />` page component from above and change it to simulate fetching data from a remote endpoint.

```tsx
// Cats.tsx

import React, {useState, useEffect} from 'react';
import {PerformanceMark} from '@shopify/react-performance';

/**
 *  Represents an entire page of our application
 */
export function Cats() {
  // simulate getting data from a remote API
  const data = useFakeRemoteData();

  return (
    <section>
      <h1>Cats</h1>
      {/* mark the page as usable immediately upon rendering */}
      <PerformanceMark stage="usable" id="cats" />
      {data && (
        <>
          {/* mark the page as *complete* once we have our data */}

>           <PerformanceMark stage="complete" id="cats" />

Reviewer 2: Should we be defaulting to just using the hook?

Author: I'm actually thinking of getting rid of the hook for this from the public API, I think you generally care where you put these and want to be able to conditionally render them, so they feel more like they want to be a component to me.

Vice versa, I don't see a reason to ever not use the hook version of `usePerformanceReport` so I'm using it in all the examples.

Reviewer 2: If we make the hook possible to use for "unfinished" cases, wouldn't that resolve the conditional rendering?

Author: I'm not sure I prefer adding more API to the hook rather than just making the component the only API for it.

Author: In a moment I'll push a big overhaul, part of which suggests only using the hook for `usePerformanceReport` and the component for `<PerformanceMark />`, let me know how you feel about it

          {data}
        </>
      )}
    </section>
  );
}

function useFakeRemoteData() {
  const [cats, setCats] = useState<string[] | null>(null);

  useEffect(() => {
    setTimeout(() => {
      setCats(['Kokusho', 'The Dude', 'Smoky', 'Moe']);
    }, 1000);
  });

  return cats;
}
```

With these changes we will see separate timing information for `usable` and `complete` in our `usePerformanceReport` output.

## API

Most APIs provided by this package are available as both hooks (ie. `useNavigationListener`), as well as components (ie. `<NavigationListener />`).

> This package also re-exports all of the API from [`@shopify/performance`](https://www.npmjs.com/package/@shopify/performance).

Reviewer: aww thats what you are going for before. Both work really, but to keep consistent I will suggest the following.

```suggestion
This package also re-exports all of the API from [`@shopify/performance`](../performance/README.md).
```

Reviewer 2: Don't link directly to the README, but agreed to link to the local package directory

#### usePerformanceReport

A custom hook which takes a url and sends a detailed breakdown of all navigations and events which took place while the page was loading. This hook should generally only be used once at the top of your component tree.

```tsx
// App.tsx
import React from 'react';
import {usePerformanceReport} from '@shopify/react-performance';
import {ProductPage} from './ProductPage';

function App() {
  usePerformanceReport('/performance-report');
  return <ProductPage />;
}
```

> To get sensical data, applications using `<PerformanceReport />` should be sure to have at least on performance mark on each top level page.

Reviewer 2: `at least one`. This is also duplicate content?

```tsx
// ProductPage
import React from 'react';
import {usePerformanceMark} from '@shopify/react-performance';

function ProductPage() {
  usePerformanceMark('complete');

  return <div>cool product page</div>;
}
```

> The data sent to the `/performance-report` url should look something like:

Reviewer 2: This is an implementation detail, since we offer abstractions on both end we should not actively show the implementation details between them
